### ISSUE TITLE
Add more pod metadata by default for filtering

### ISSUE THREAD

**Issue by enote-kane** [falcosecurity/falco#2791]
**Motivation**

In order to filter all events for a newly deployed Helm Chart, I want to have additional `output_fields` available.

**Feature**

I would like to see the following metadata attached:
* pod labels
* container name

**Alternatives**

None considered since I didn't find documentation how to add this information through custom rules.

**Comment by Andreagit97** [comment]
ei thank you for reporting!

> container name

This should be available as `container.name` -> https://falco.org/docs/reference/rules/supported-fields/#field-class-container

> pod labels

For this one we should double check with @alacuku if can be easily extracted without extra overhead in the new k8s client

**Comment by alacuku** [comment]
Hi @enote-kane, if you need more fields to be outputted in the default rules the only way would be to patch the rules and add them. Please have a look here:
* https://falco.org/docs/rules/appending/ 
* https://falco.org/blog/gitops-your-falco-rules/

**Comment by incertum** [comment]
In addition, please check out the latest official docs to understand the placeholder `%container.info` in the upstream rules

https://falco.org/docs/rules/style-guide/
https://falco.org/docs/outputs/formatting/

And as Andrea pointed out we expose many fields for container and k8s. Many of the k8s fields are extracted from the container runtime.

https://falco.org/docs/reference/rules/supported-fields/#field-class-container
https://falco.org/docs/reference/rules/supported-fields/#field-class-k8s

**Comment by enote-kane** [comment]
@alacuku @incertum  Thanx for the info, but none of it talks about whether (and how) I can make use of those fields for example in the `output_fields` map (they only talk about how to use them for the `output` message). Since I don't see any other field in the rules that could be used for that. The default rules do not seem to specify any and it appears it will produce "default" `output_fields`.

My actual goal here would be to easily filter in the web UI for all events related to service `nginx-ingress-controller`.

So referring to the docs, the only example I see is the one for JSON log format: https://falco.org/docs/outputs/channels/
```json
{
    "hostname": "falco-xczjd",
    "output": "13:44:05.478445995: Critical A shell was spawned in a container with an attached terminal (user=root user_loginuid=-1 k8s.ns=default k8s.pod=kubecon container=ee97d9c4186f shell=sh parent=runc cmdline=sh -c clear; (bash || ash || sh) terminal=34816 container_id=ee97d9c4186f image=docker.io/library/alpine)",
    "priority": "Critical",
    "rule": "Terminal shell in container",
    "source": "syscall",
    "tags": [
        "container",
        "mitre_execution",
        "shell"
    ],
    "time": "2023-05-25T13:44:05.478445995Z",
    "output_fields": {  <<< add custom entries here per rule?
        "container.id": "ee97d9c4186f",
        "container.image.repository": "docker.io/library/alpine",
        "evt.time": 1685022245478445995,
        "k8s.ns.name": "default",
        "k8s.pod.name": "kubecon",
        "proc.cmdline": "sh -c clear; (bash || ash || sh)",
        "proc.name": "sh",
        "proc.pname": "runc",
        "proc.tty": 34816,
        "user.loginuid": -1,
        "user.name": "root"
    }
}
```

So how exactly do I adjust a rule to get a custom entry in that above `output_fields` object?

**Comment by incertum** [comment]
Here are a few clarifying questions:

- What field(s) do you want per rule? `<<< add custom entries here per rule?` Is it an existing field listed on the site you want? https://falco.org/docs/reference/rules/supported-fields/ In that case simple add the field(s) you want to the output yaml property. You can do that manually (add the filed with the % preceding the field name) or use the append feature as @alacuku mentioned which however will be reworked in future releases
```
output: System user ran an interactive command (pod_labels=%k8s.pod.labels evt_type=%evt.type user=%user.name user_uid=%user.uid user_loginuid=%user.loginuid process=%proc.name proc_exepath=%proc.exepath parent=%proc.pname command=%proc.cmdline terminal=%proc.tty exe_flags=%evt.arg.flags %container.info)
```
- Or do you want a complete new field not listed on the site? In that case the Falco libs source would need to be patched to feature this new field

**Comment by enote-kane** [comment]
@incertum Thanx for the additional questions.

OK, so does Falco scan the `output` field message template and add any `%...` keys to the `output_fields` map? If that's the case, then I think I should be good (never tried it).

In that case, I am wondering whether Falco will make sure that supported field `k8s.svc.name` will be made available to events that originate from pods what are part of the service (through their endpoints).

I would have the use-case of defining a new `output_field` key, but so far more in the sense of renaming an supported field (e.g. `k8s.svc.name` → `service_name`).

In the future, I would also like to add completely custom fields based on other labels, e.g. add output field `zone`, defined as the value of label `topology.kubernetes.io/zone` from the node where the pod is running at. As other tools also add custom labels, having an option to add fields (not only in the output, but also usable for the rule itself) would greatly increase flexibility of Falco, depending on the specific setup of a cluster.

Documentation is not clear at all how the `output_fields` map is generated. I just assume that this is used to generate the potential columns/fields in the UI as well. If this assumption is wrong, then please correct me.

**Comment by incertum** [comment]
> OK, so does Falco scan the output field message template and add any %... keys to the output_fields map? If that's the case, then I think I should be good (never tried it).

Yes curious if you have time to read all of the Falco rules documentation, how can we make this more clear? https://falco.org/docs/rules/

> In that case, I am wondering whether Falco will make sure that supported field k8s.svc.name will be made available to events that originate from pods what are part of the service (through their endpoints).

Same question, how can we make docs clearer https://falco.org/docs/reference/rules/supported-fields/#field-class-k8s

```
Kubernetes related context. When configured to fetch from the API server, all fields are available. Otherwise, only the k8s.pod.* and k8s.ns.name fields are populated with data gathered from the container runtime.
```

Suggest reading also all relevant sections around that in https://github.com/falcosecurity/falco/blob/master/falco.yaml and the cmd args help prints


> In the future, I would also like to add completely custom fields based on other labels, e.g. add output field zone, defined as the value of label topology.kubernetes.io/zone from the node where the pod is running at. As other tools also add custom labels, having an option to add fields (not only in the output, but also usable for the rule itself) would greatly increase flexibility of Falco, depending on the specific setup of a cluster.

We have had similar requests, meanwhile please use `evt.hostname` (read desc in the output fields page), it should contain cluster name when set to the right kube node information and you can customize it through setting the FALCO_HOSTNAME env to what you want on each host.

User defined completely custom fields are not on the roadmap, perhaps open a new feature request for that?

**Comment by incertum** [comment]
Thank you @enote-kane. I have created a follow up ticket in the falco website repo. We will try our best to further improve docs.
